File: tests/fixtures/structured-types/recursive-installed.py
import concurrent.futures
import dataclasses
import inspect
import json
import math
import sys
import time
import typing
import lean_recursive as lb

checks = rejected = 0


def check(condition):
    global checks
    assert condition
    checks += 1


def rejects(error_type, call, status=None):
    global rejected
    try:
        call()
    except error_type as error:
        check(status is None or error.status == status)
        rejected += 1
    else:
        raise AssertionError(f"Expected {error_type}")


def scalar(**updates):
    return dataclasses.replace(lb.Scalars(
        None, True, 255, 65535, (1 << 32) - 1, (1 << 64) - 1,
        -128, -32768, -(1 << 31), -(1 << 63), (1 << 128) + 1,
        -((1 << 128) + 1), 1.5, -2.25, "A\0🌱", b"\0\xff\x01", "🌱",
        (1 << 32) - 1, -(1 << 31)), **updates)


value = scalar()
check(lb.inspect(value))  # This predicate interprets all nineteen fields in Lean.
check(lb.scalars(value) == value)
check(lb.word_max((1 << 64) - 1))
check(lb.signed_min(-(1 << 63)))
for change in [dict(natural=(1 << 1000) + 7, integer=-((1 << 1000) + 7)),
               dict(natural=0, integer=0, text="", bytes_=b""),
               dict(text="\0a🌿\0", bytes_=bytes(range(256))),
               dict(f32=-math.inf, f64=math.inf)]:
    item = scalar(**change)
    check(lb.scalars(item) == item)
special = lb.scalars(scalar(f32=math.nan, f64=-0.0))
check(math.isnan(special.f32) and math.copysign(1, special.f64) == -1)
check(lb.scalars(scalar(f32=1 + 2**-24)).f32 == 1)
for field, bits, signed in [("u8", 8, False), ("u16", 16, False), ("u32", 32, False),
                            ("u64", 64, False), ("i8", 8, True), ("i16", 16, True),
                            ("i32", 32, True), ("i64", 64, True),
                            ("word", 64, False), ("signed_word", 64, True)]:
    minimum = -(1 << (bits - 1)) if signed else 0
    maximum = (1 << (bits - int(signed))) - 1
    for number in [minimum, maximum]:
        check(getattr(lb.scalars(scalar(**{field: number})), field) == number)
    for number in [minimum - 1, maximum + 1]:
        invalid = scalar(**{field: number})
        rejects(ValueError, lambda: lb.scalars(invalid))
    for number in [True, 1.0, "1"]:
        invalid = scalar(**{field: number})
        rejects(TypeError, lambda: lb.scalars(invalid))
for field, bad in [("unit", 0), ("bool_", 1), ("natural", True), ("integer", 1.0),
                   ("f32", 1), ("f64", True), ("bytes_", bytearray(b"x")), ("text", b"x"),
                   ("char_", 65)]:
    invalid = scalar(**{field: bad})
    rejects(TypeError, lambda: lb.scalars(invalid))
for field, bad in [("natural", -1), ("char_", ""), ("char_", "ab"), ("char_", "\ud800"),
                   ("text", "\udfff")]:
    invalid = scalar(**{field: bad})
    rejects(ValueError, lambda: lb.scalars(invalid))
tree = lb.TreeBranch((lb.TreeLeaf(value), lb.TreeBranch(())))
check(lb.tree(tree) == tree)
check(lb.empty() == lb.TreeBranch(()))
check(lb.join_trees(tree, tree) == lb.TreeBranch((tree, tree)))
check(lb.forest([tree] * 512) == (tree,) * 512)
check(lb.forest(()) == ())
input_list = [lb.TreeBranch([])]
copied = lb.tree(lb.TreeBranch(input_list))
input_list[0].children.append(lb.TreeLeaf(value))
check(copied == lb.TreeBranch((lb.TreeBranch(()),)))
alias_copies = lb.forest([tree, tree])
check(alias_copies[0] is not alias_copies[1])
for marker in [None, lb.Some(None), lb.Some(lb.Some(None))]:
    for outcome in [lb.Ok((tree, tree)), lb.Err("error\0🌱")]:
        for fallback in [None, lb.Some(tree)]:
            envelope = lb.Envelope(tree, ((), (tree,)), fallback, outcome, marker)
            check(lb.envelope(envelope) == envelope)
left = lb.LeftTreeNext(lb.RightTreeMany((lb.LeftTreeLeaf(9),)))
check(lb.left(left) == left)
check(lb.right(lb.RightTreeMany((left,))) == lb.RightTreeMany((left,)))
spine = lb.SpineLeaf(41)
for _ in range(127):
    spine = lb.SpineNext(spine)
a, b = spine, lb.spine(spine)
for _ in range(127):
    check(a is not b)
    a, b = a.value, b.value
check(a.value == b.value == 41)
rejects(lb.LeanBridgeError, lambda: lb.grow(spine), 2)
rejects(ValueError, lambda: lb.spine(lb.SpineNext(spine)))
check(lb.grow(lb.SpineLeaf(7)) == lb.SpineNext(lb.SpineLeaf(7)))
wide = lb.WideNext(**{f"field{i}": i for i in range(255)}, child=lb.WideLeaf(17))
check(lb.wide(wide) == wide)
for marker in [lb.MarkerEmpty(), lb.MarkerUnit(None), lb.MarkerNext(lb.MarkerEmpty())]:
    check(lb.marker(marker) == marker)
check(lb.empty_record(lb.EmptyRecord()) == lb.EmptyRecord())
check(lb.units([None] * 123) == (None,) * 123)
rejects(ValueError, lambda: lb.never(object()))
rejects(TypeError, lambda: lb.spine(lb.MarkerEmpty()))
rejects(TypeError, lambda: lb.tree({"kind": "branch", "children": []}))
rejects(TypeError, lambda: lb.envelope(dataclasses.replace(envelope, marker=True)))
rejects(TypeError, lambda: lb.join_trees(tree, lb.TreeLeaf(scalar(u32=True))))
rejects(dataclasses.FrozenInstanceError, lambda: setattr(tree, "children", ()))


class FakeLeaf(lb.SpineLeaf):
    pass


rejects(TypeError, lambda: lb.spine(FakeLeaf(1)))
cyclic = []
cyclic_tree = lb.TreeBranch(cyclic)
cyclic.append(cyclic_tree)
rejects(ValueError, lambda: lb.tree(cyclic_tree))
cyclic.clear()
rejects(ValueError, lambda: lb.forest([lb.TreeBranch(())] * 262144))
rejects(ValueError, lambda: lb.scalars(scalar(text="x" * (16 * 1024 * 1024))))
check(lb.tree(tree) == tree)
check(lb.inspect(value))


def threaded(index):
    own = lb.SpineNext(lb.SpineLeaf(index))
    for _ in range(16):
        assert lb.spine(own) == own
    return index


with concurrent.futures.ThreadPoolExecutor(max_workers=4) as pool:
    check(list(pool.map(threaded, range(16))) == list(range(16)))
functions = [name for name in lb.__all__ if inspect.isfunction(getattr(lb, name))]
check(len(functions) == 18)
started = time.monotonic()
for name in functions:
    hints = typing.get_type_hints(getattr(lb, name))
    check("return" in hints and len(repr(hints)) < 10000)
    check("Any" not in repr(hints))
for record in [lb.Envelope, lb.TreeBranch, lb.WideNext, lb.Scalars]:
    check(len(repr(typing.get_type_hints(record))) < 10000)
check(time.monotonic() - started < 2)
parameter = next(iter(inspect.signature(lb.spine).parameters))
check(lb.spine(**{parameter: lb.SpineLeaf(19)}) == lb.SpineLeaf(19))
print(json.dumps({"checks": checks, "rejected": rejected, "functions": len(functions),
                  "threads": 4, "threadedCalls": 256, "python": sys.version.split()[0]}))

File: tests/fixtures/structured-types/recursive-installed.rs
// Independent safe Rust caller. Only the installed public crate is imported.
use recursive_api as api;
use api::{BigInt, BigUint, EmptyRecord, Envelope, LeftTree, Marker, RightTree, Scalars, Spine, Tree, Wide};
thread_local! { static CHECKS: std::cell::Cell<usize> = const { std::cell::Cell::new(0) }; }
macro_rules! check { ($expression:expr) => {{ CHECKS.with(|n| n.set(n.get() + 1)); assert!($expression); }}; }
fn payload() -> Scalars {
    let natural = (BigUint::from(1u8) << 128usize) + 1u8;
    Scalars { unit: (), bool_: true, u8: u8::MAX, u16: u16::MAX, u32: u32::MAX, u64: u64::MAX,
        i8: i8::MIN, i16: i16::MIN, i32: i32::MIN, i64: i64::MIN,
        integer: -BigInt::from(natural.clone()), natural, f32: 1.5, f64: -2.25,
        text: "A\0🌱".into(), bytes: vec![0, 255, 1], char_: '🌱', word: u32::MAX.into(), signed_word: i32::MIN.into() }
}
fn main() -> Result<(), api::Error> {
    let mut too_deep = Tree::Branch { children: vec![] };
    for _ in 0..130 { too_deep = Tree::Branch { children: vec![too_deep] }; }
    let leaf = Tree::Leaf { payload: payload() };
    check!(api::join_trees(&leaf, &too_deep) == Err(api::Error::Limit));
    let scalar = payload();
    check!(api::inspect(&scalar)?); check!(api::scalars(&scalar)? == scalar);
    check!(api::word_max(u64::MAX)?); check!(api::signed_min(i64::MIN)?);
    let mut large = scalar.clone(); large.natural = (BigUint::from(1u8) << 1000usize) + 7u8;
    large.integer = -BigInt::from(large.natural.clone());
    check!(api::scalars(&large)? == large);
    let mut empty = scalar.clone(); empty.natural = 0u8.into(); empty.integer = 0u8.into(); empty.text.clear(); empty.bytes.clear();
    check!(api::scalars(&empty)? == empty);
    let mut special = scalar.clone(); special.f32 = f32::NAN; special.f64 = -0.0;
    let copy = api::scalars(&special)?; check!(copy.f32.is_nan()); check!(copy.f64.to_bits() == (-0.0f64).to_bits());
    special.f32 = f32::NEG_INFINITY; special.f64 = f64::INFINITY;
    check!(api::scalars(&special)? == special);

    let tree: api::TreeAlias = Tree::Branch { children: vec![leaf.clone(), Tree::Branch { children: vec![] }] };
    check!(api::tree(&tree)? == tree);
    check!(api::join_trees(&tree, &leaf)? == Tree::Branch { children: vec![tree.clone(), leaf.clone()] });
    check!(api::empty()? == Tree::Branch { children: vec![] });
    let forest: api::Forest = vec![tree.clone(); 512]; check!(api::forest(&forest)? == forest);
    let mut envelope = Envelope { tree: tree.clone(), alternatives: vec![vec![], vec![tree.clone()]], fallback: Some(leaf.clone()),
        outcome: Box::new(Ok((tree.clone(), leaf.clone()))), marker: None };
    for marker in [None, Some(None), Some(Some(()))] {
        envelope.marker = marker; check!(api::envelope(&envelope)? == envelope);
        let mut error = envelope.clone(); error.outcome = Box::new(Err("error\0🌿".into())); error.fallback = None;
        check!(api::envelope(&error)? == error);
    }
    let left = LeftTree::Next { right: RightTree::Many { lefts: vec![LeftTree::Leaf { value: 9 }] } };
    check!(api::left(&left)? == left);
    let right = RightTree::Many { lefts: vec![left] }; check!(api::right(&right)? == right);
    let mut spine = Spine::Leaf { value: 41 };
    for _ in 0..127 { spine = Spine::Next { value: Box::new(spine) }; }
    let mut copy = api::spine(&spine)?; check!(copy == spine);
    let (mut a, mut b) = (&spine, &mut copy);
    for _ in 0..127 {
        check!(!std::ptr::eq(a, b));
        a = match a { Spine::Next { value } => value, _ => panic!("missing input child") };
        b = match b { Spine::Next { value } => value, _ => panic!("missing copied child") };
    }
    match b { Spine::Leaf { value } => *value = 42, _ => panic!("missing leaf") };
    check!(copy != spine);
    check!(api::grow(&spine) == Err(api::Error::Limit));
    check!(api::grow(&Spine::Leaf { value: 7 })? == Spine::Next { value: Box::new(Spine::Leaf { value: 7 }) });
    let wide = Wide::Next { __WIDE_FIELDS__ child: Box::new(Wide::Leaf { value: 17 }) };
    check!(api::wide(&wide)? == wide);
    for marker in [Marker::Empty, Marker::Unit { value: () }, Marker::Next { value: Box::new(Marker::Empty) }] {
        check!(api::marker(&marker)? == marker);
    }
    check!(api::empty_record(&EmptyRecord {})? == EmptyRecord {});
    check!(api::units(&[(); 123])?.len() == 123);
    check!(api::units(&vec![(); 262_144]) == Err(api::Error::Limit));
    for _ in 0..100 { check!(api::envelope(&envelope)? == envelope); }
    std::thread::scope(|scope| {
        for _ in 0..4 { scope.spawn(|| { for _ in 0..25 { assert!(api::inspect(&scalar).unwrap()); } }); }
    });
    let mappings = std::fs::read_to_string("/proc/self/maps").unwrap();
    let paths: std::collections::BTreeSet<_> = mappings.lines().filter_map(|line|
        line.split_whitespace().find(|word| word.starts_with("/tmp/lean-bridge-rust-assets-"))).collect();
    check!(paths.len() == 4);
    for path in paths { check!(!std::path::Path::new(path).parent().unwrap().exists()); }
    println!("recursive-installed-ok:{}", CHECKS.with(|n| n.get()) + 100);
    Ok(())
}
